File: Ch03/WordList/WordList/Persistence.m
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

#import "Persistence.h"

@implementation Persistence

- (id)init {
  self = [super init];
  if (self != nil) {
    // Initialize the managed object model
    NSURL *modelURL = [[NSBundle mainBundle] URLForResource:@"WordList" withExtension:@"momd"];
    _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:modelURL];
    
    // Initialize the persistent store coordinator
    NSURL *storeURL = [[self applicationDocumentsDirectory] URLByAppendingPathComponent:@"WordList.sqlite"];
    
    NSError *error = nil;
    _persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:self.managedObjectModel];
    if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType
                                                   configuration:nil
                                                             URL:storeURL
                                                         options:nil
                                                           error:&error]) {
      NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
      abort();
    }
    
    // Initialize the managed object context
    _managedObjectContext = [[NSManagedObjectContext alloc] init];
    [_managedObjectContext setPersistentStoreCoordinator:self.persistentStoreCoordinator];
  }
  return self;
}

- (void)loadWordList:(NSString *)wordList {
  // Delete all the existing words and categories
  [self deleteAllObjectsForEntityWithName:@"Word"];
  [self deleteAllObjectsForEntityWithName:@"WordCategory"];
  
  // Create the categories
  NSMutableDictionary *wordCategories = [NSMutableDictionary dictionaryWithCapacity:26];
  for (char c = 'a'; c <= 'z'; c++) {
    NSString *firstLetter = [NSString stringWithFormat:@"%c", c];
    NSManagedObject *wordCategory = [NSEntityDescription insertNewObjectForEntityForName:@"WordCategory" inManagedObjectContext:self.managedObjectContext];
    [wordCategory setValue:firstLetter forKey:@"firstLetter"];
    [wordCategories setValue:wordCategory forKey:firstLetter];
    NSLog(@"Added category '%@'", firstLetter);
  }
  
  // Add the words from the list
  NSUInteger wordsAdded = 0;
  NSArray *newWords = [wordList componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  for (NSString *word in newWords) {
    if (word.length > 0) {
      NSManagedObject *object = [NSEntityDescription insertNewObjectForEntityForName:@"Word" inManagedObjectContext:self.managedObjectContext];
      [object setValue:word forKey:@"text"];
      [object setValue:[NSNumber numberWithInteger:word.length] forKey:@"length"];
      [object setValue:[wordCategories valueForKey:[word substringToIndex:1]] forKey:@"wordCategory"];
      ++wordsAdded;
      if (wordsAdded % 100 == 0)
        NSLog(@"Added %lu words", wordsAdded);
    }
  }
  NSLog(@"Added %lu words", wordsAdded);
  [self saveContext];
  NSLog(@"Context saved");
}

- (NSString *)statistics {
  NSMutableString *string = [[NSMutableString alloc] init];
  [string appendString:[self wordCount]];
  for (char c = 'a'; c <= 'z'; c++) {
    [string appendString:[self wordCountForCategory:[NSString stringWithFormat:@"%c", c]]];
  }
  [string appendString:[self zyzzyvasUsingQueryLanguage]];
  [string appendString:[self zyzzyvasUsingNSExpression]];
  [string appendString:[self wordCountForRange:NSMakeRange(20, 25)]];
  [string appendString:[self endsWithGryWords]];
  [string appendString:[self anyWordContainsZ]];
  [string appendString:[self caseInsensitiveFetch:@"qiviut"]];
  [string appendString:[self twentyLetterWordsEndingInIng]];
  [string appendString:[self highCountCategories]];
  [string appendString:[self averageWordLengthFromCollection]];
  [string appendString:[self averageWordLengthFromExpressionDescription]];
  [string appendString:[self longestWords]];
  [string appendString:[self wordCategoriesWith25LetterWords]];
  [string appendString:[self zWords]];
  return string;
}

- (NSString *)wordCount {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  NSUInteger count = [self.managedObjectContext countForFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Word Count: %lu\n", count];
}

- (NSString *)wordCountForCategory:(NSString *)firstLetter {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = [NSPredicate predicateWithFormat:@"wordCategory.firstLetter = %@", firstLetter];
  NSUInteger count = [self.managedObjectContext countForFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Words beginning with %@: %lu\n", firstLetter, count];
}

- (NSString *)zyzzyvasUsingQueryLanguage {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = [NSPredicate predicateWithFormat:@"text = 'zyzzyvas'"];
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return words.count == 0 ? @"" : [NSString stringWithFormat:@"%@\n", [words[0] valueForKey:@"text"]];
}

- (NSString *)zyzzyvasUsingNSExpression {
  NSExpression *expressionText = [NSExpression expressionForKeyPath:@"text"];
  NSExpression *expressionZyzzyvas = [NSExpression expressionForConstantValue:@"zyzzyvas"];
  NSPredicate *predicate = [NSComparisonPredicate predicateWithLeftExpression:expressionText
                                                              rightExpression:expressionZyzzyvas
                                                                     modifier:NSDirectPredicateModifier
                                                                         type:NSEqualToPredicateOperatorType
                                                                      options:0];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = predicate;
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return words.count == 0 ? @"" : [NSString stringWithFormat:@"%@\n", [words[0] valueForKey:@"text"]];
}

- (NSString *)wordCountForRange:(NSRange)range {
  NSExpression *length = [NSExpression expressionForKeyPath:@"length"];
  NSExpression *lower = [NSExpression expressionForConstantValue:@(range.location)];
  NSExpression *upper = [NSExpression expressionForConstantValue:@(range.length)];
  NSExpression *expr = [NSExpression expressionForAggregate:@[lower, upper]];
  NSPredicate *predicate = [NSComparisonPredicate predicateWithLeftExpression:length
                                                              rightExpression:expr
                                                                     modifier:NSDirectPredicateModifier
                                                                         type:NSBetweenPredicateOperatorType
                                                                      options:0];
  NSLog(@"Aggregate Predicate: %@", [predicate predicateFormat]);
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = predicate;
  NSUInteger count = [self.managedObjectContext countForFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"%lu-%lu letter words: %lu\n", range.location, range.length, count];
}

- (NSString *)endsWithGryWords {
  NSExpression *text = [NSExpression expressionForKeyPath:@"text"];
  NSExpression *gry = [NSExpression expressionForConstantValue:@"gry"];
  NSPredicate *predicate = [NSComparisonPredicate predicateWithLeftExpression:text
                                                              rightExpression:gry
                                                                     modifier:NSDirectPredicateModifier
                                                                         type:NSEndsWithPredicateOperatorType
                                                                      options:0];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = predicate;
  NSArray *gryWords = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"-gry words: %@\n",
          [[gryWords valueForKey:@"text"] componentsJoinedByString:@","]];
}

- (NSString *)anyWordContainsZ {
  NSExpression *text = [NSExpression expressionForKeyPath:@"words.text"];
  NSExpression *z = [NSExpression expressionForConstantValue:@"z"];
  NSPredicate *predicate = [NSComparisonPredicate predicateWithLeftExpression:text
                                                              rightExpression:z
                                                                     modifier:NSAnyPredicateModifier
                                                                         type:NSContainsPredicateOperatorType
                                                                      options:0];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"WordCategory"];
  fetchRequest.predicate = predicate;
  NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"ANY: %@\n",
          [[categories valueForKey:@"firstLetter"] componentsJoinedByString:@","]];
}

- (NSString *)caseInsensitiveFetch:(NSString *)word {
  NSExpression *text = [NSExpression expressionForKeyPath:@"text"];
  NSExpression *allCapsWord = [NSExpression expressionForConstantValue:[word uppercaseString]];
  NSPredicate *predicate = [NSComparisonPredicate predicateWithLeftExpression:text
                                                              rightExpression:allCapsWord
                                                                     modifier:NSDirectPredicateModifier
                                                                         type:NSEqualToPredicateOperatorType
                                                                      options:NSCaseInsensitivePredicateOption];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = predicate;
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"%@\n", words.count == 0 ? @"" : [words[0] valueForKey:@"text"]];
}

- (NSString *)twentyLetterWordsEndingInIng {
  // Create a predicate that compares length to 20
  NSExpression *length = [NSExpression expressionForKeyPath:@"length"];
  NSExpression *twenty = [NSExpression expressionForConstantValue:@20];
  NSPredicate *predicateLength = [NSComparisonPredicate predicateWithLeftExpression:length
                                                                    rightExpression:twenty
                                                                           modifier:NSDirectPredicateModifier
                                                                               type:NSEqualToPredicateOperatorType
                                                                            options:0];
  
  // Create a predicate that compares text to "ends with -ing"
  NSExpression *text = [NSExpression expressionForKeyPath:@"text"];
  NSExpression *ing = [NSExpression expressionForConstantValue:@"ing"];
  NSPredicate *predicateIng = [NSComparisonPredicate predicateWithLeftExpression:text
                                                                 rightExpression:ing
                                                                        modifier:NSDirectPredicateModifier
                                                                            type:NSEndsWithPredicateOperatorType
                                                                         options:0];
  
  // Combine the predicates
  NSPredicate *predicate = [NSCompoundPredicate andPredicateWithSubpredicates:@[predicateLength, predicateIng]];
  NSLog(@"Compound predicate: %@", [predicate predicateFormat]);
  
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.predicate = predicate;
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"%@\n", [[words valueForKey:@"text"] componentsJoinedByString:@","]];
}

- (NSString *)highCountCategories {
  NSPredicate *predicate = [NSPredicate predicateWithFormat:@"words.@count > %d", 10000];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"WordCategory"];
  fetchRequest.predicate = predicate;
  NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"High count categories: %@\n",
          [[categories valueForKey:@"firstLetter"] componentsJoinedByString:@","]];
}

- (NSString *)averageWordLengthFromCollection {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Average from collection: %.2f\n",
          [[words valueForKeyPath:@"@avg.length"] floatValue]];
}

- (NSString *)averageWordLengthFromExpressionDescription {
  NSExpression *length = [NSExpression expressionForKeyPath:@"length"];
  NSExpression *average = [NSExpression expressionForFunction:@"average:" arguments:@[length]];
  
  NSExpressionDescription *averageDescription = [[NSExpressionDescription alloc] init];
  averageDescription.name = @"average";
  averageDescription.expression = average;
  averageDescription.expressionResultType = NSFloatAttributeType;
  
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  fetchRequest.propertiesToFetch = @[averageDescription];
  fetchRequest.resultType = NSDictionaryResultType;
  NSArray *results = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Average from expression description: %.2f\n",
          [[results[0] valueForKey:@"average"] floatValue]];
}

- (NSString *)longestWords {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  NSPredicate *predicate = [NSPredicate predicateWithFormat:@"length = max:(length)"];
  NSLog(@"Predicate: %@", [predicate predicateFormat]);
  fetchRequest.predicate = predicate;
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Longest words: %@\n",
          [[words valueForKey:@"text"] componentsJoinedByString:@","]];
}

- (NSString *)wordCategoriesWith25LetterWords {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"WordCategory"];
  NSPredicate *predicate = [NSPredicate predicateWithFormat:@"SUBQUERY(words, $x, $x.length = 25).@count > 0"];
  fetchRequest.predicate = predicate;
  NSLog(@"Subquery Predicate: %@", [predicate predicateFormat]);
  NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Categories with 25-letter words: %@\n",
          [[categories valueForKey:@"firstLetter"] componentsJoinedByString:@","]];
}

- (NSString *)zWords {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Word"];
  NSPredicate *predicate = [NSPredicate predicateWithFormat:@"text BEGINSWITH 'z'"];
  NSSortDescriptor *lengthSort = [NSSortDescriptor sortDescriptorWithKey:@"length" ascending:YES];
  NSSortDescriptor *alphaSort = [NSSortDescriptor sortDescriptorWithKey:@"text" ascending:NO];
  
  fetchRequest.predicate = predicate;
  fetchRequest.sortDescriptors = @[lengthSort, alphaSort];
  
  NSLog(@"Predicate: %@", predicate);
  NSArray *words = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  return [NSString stringWithFormat:@"Z words:\n%@\n",
          [[words valueForKey:@"text"] componentsJoinedByString:@"\n"]];
}

#pragma mark - Helper Methods

- (void)deleteAllObjectsForEntityWithName:(NSString *)name {
  NSLog(@"Deleting all objects in entity %@", name);
  
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:name];
  fetchRequest.resultType = NSManagedObjectIDResultType;
  
  NSArray *objectIDs = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  for (NSManagedObjectID *objectID in objectIDs) {
    [self.managedObjectContext deleteObject:[self.managedObjectContext objectWithID:objectID]];
  }
  
  [self saveContext];
  
  NSLog(@"All objects in entity %@ deleted", name);
}

- (void)saveContext {
  NSError *error = nil;
  if ([self.managedObjectContext hasChanges] && ![self.managedObjectContext save:&error]) {
    NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
    abort();
  }
}

- (NSURL *)applicationDocumentsDirectory {
  return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

@end

File: Ch03/WordListSwift/WordListSwift/Persistence.swift
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

import Foundation
import CoreData

class Persistence: NSObject {
    var managedObjectContext: NSManagedObjectContext? = {
        // Initialize the managed object model
        let modelURL = NSBundle.mainBundle().URLForResource("WordListSwift", withExtension: "momd")
        let managedObjectModel = NSManagedObjectModel(contentsOfURL: modelURL!)
        
        // Initialize the persistent store coordinator
        let storeURL = Persistence.applicationDocumentsDirectory.URLByAppendingPathComponent("WordListSwift.sqlite")
        var error: NSError? = nil
        let persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel!)

        if(persistentStoreCoordinator.addPersistentStoreWithType(NSSQLiteStoreType, configuration: nil, URL: storeURL, options: nil, error: &error) == nil) {
            abort()
        }
        
        var managedObjectContext = NSManagedObjectContext()
        managedObjectContext.persistentStoreCoordinator = persistentStoreCoordinator
        
        return managedObjectContext
    }()
    
    
    func saveContext() {
        var error: NSError? = nil
        if let managedObjectContext = self.managedObjectContext {
            if managedObjectContext.hasChanges && !managedObjectContext.save(&error) {
                abort()
            }
        }
    }
  
    class var applicationDocumentsDirectory: NSURL {
        let urls = NSFileManager.defaultManager().URLsForDirectory(.DocumentDirectory, inDomains: .UserDomainMask)
        return urls[urls.endIndex-1] as NSURL
    }
    
    func deleteAllObjectsForEntityWithName(name: String) {
        println("Deleting all objects in entity \(name)")
        var fetchRequest = NSFetchRequest(entityName: name)
        fetchRequest.resultType = .ManagedObjectIDResultType
        
        if let managedObjectContext = managedObjectContext {
            var error: NSError? = nil
            let objectIDs = managedObjectContext.executeFetchRequest(fetchRequest, error: &error)
            for objectID in objectIDs! {
                managedObjectContext.deleteObject(managedObjectContext.objectWithID(objectID as NSManagedObjectID))
            }
            
            saveContext()
            
            println("All objects in entity \(name) deleted")
        }
    }
    
    func loadWordList(wordList: String) {
        // Delete all the existing words and categories
        deleteAllObjectsForEntityWithName("Word")
        deleteAllObjectsForEntityWithName("WordCategory")
        
        // Create the categories
        var wordCategories = NSMutableDictionary(capacity: 26)
        for c in "abcdefghijklmnopqrstuvwxyz" {
            let firstLetter = "\(c)"
            var wordCategory: AnyObject! = NSEntityDescription.insertNewObjectForEntityForName("WordCategory", inManagedObjectContext: self.managedObjectContext!)
            wordCategory.setValue(firstLetter, forKey: "firstLetter")
            wordCategories.setValue(wordCategory, forKey: firstLetter)
            println("Added category '\(firstLetter)'")
        }
        
        // Add the words from the list
        var wordsAdded = 0
        let newWords = wordList.componentsSeparatedByCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
        for word in newWords {
            if countElements(word) > 0 {
                var object: AnyObject! = NSEntityDescription.insertNewObjectForEntityForName("Word", inManagedObjectContext: self.managedObjectContext!)
                object.setValue(word, forKey: "text")
                object.setValue(countElements(word), forKey: "length")
                object.setValue(wordCategories.valueForKey((word as NSString).substringToIndex(1)), forKey: "wordCategory")
                wordsAdded++
                if wordsAdded % 100 == 0 {
                    println("Added \(wordsAdded) words")
                    saveContext()
                    println("Context saved")
                }
            }
        }
        saveContext()
        println("Context saved")
    }
    
    func wordCount() -> String {
        let fetchRequest = NSFetchRequest(entityName: "Word")
        var error: NSError? = nil
        let count = self.managedObjectContext!.countForFetchRequest(fetchRequest, error: &error)
        return "Word Count: \(count)\n"
    }
    
    func statistics() -> String {
        var string = NSMutableString()

        string.appendString(wordCount())
        
        for c in "abcdefghijklmnopqrstuvwxyz" {
            string.appendString(wordCountForCategory("\(c)"))
        }

        string.appendString(zyzzyvasUsingQueryLanguage())
        string.appendString(zyzzyvasUsingNSExpression())
        string.appendString(wordCountForRange(NSMakeRange(20, 25)))
        string.appendString(endsWithGryWords())
        string.appendString(anyWordContainsZ())
        string.appendString(caseInsensitiveFetch("qiviut"))
        string.appendString(twentyLetterWordsEndingInIng())
        string.appendString(highCountCategories())
        string.appendString(averageWordLengthFromCollection())
        string.appendString(averageWordLengthFromExpressionDescription())
        string.appendString(longestWords())
        string.appendString(wordCategoriesWith25LetterWords())
        string.appendString(zWords())

        return string
    }
    
    func wordCountForCategory(firstLetter: String) -> String {
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = NSPredicate(format: "wordCategory.firstLetter = %@", argumentArray: [firstLetter])
        
        var error: NSError? = nil
        let count = self.managedObjectContext!.countForFetchRequest(fetchRequest, error: &error)
        
        return "Words beginning with \(firstLetter): \(count)\n"
    }
    
    func zyzzyvasUsingQueryLanguage() -> String {
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = NSPredicate(format: "text = 'zyzzyvas'", argumentArray: [])

        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            if words.isEmpty { return "" }
            else {
                let word: AnyObject! = words[0].valueForKey("text")
                return "\(word)\n"
            }
        }
        else { return "" }
    }
    
    func zyzzyvasUsingNSExpression() -> String {
        let expressionText = NSExpression(forKeyPath: "text")
        let expressionZyzzyvas = NSExpression(forConstantValue: "zyzzyvas")
        let predicate = NSComparisonPredicate(leftExpression: expressionText, rightExpression: expressionZyzzyvas, modifier: .DirectPredicateModifier, type: .EqualToPredicateOperatorType, options: nil)
        
        println("Predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = predicate
        
        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            if words.isEmpty { return "" }
            else {
                let word: AnyObject! = words[0].valueForKey("text")
                return "\(word)\n"
            }
        }
        else { return "" }
    }
    
    func wordCountForRange(range: NSRange) -> String {
        let length = NSExpression(forKeyPath: "length")
        let lower = NSExpression(forConstantValue: range.location)
        let upper = NSExpression(forConstantValue: range.length)
        let expr = NSExpression(forAggregate: [lower, upper])
        let predicate = NSComparisonPredicate(leftExpression: length, rightExpression: expr, modifier: .DirectPredicateModifier, type: .BetweenPredicateOperatorType, options: nil)
        
        println("Aggregate predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = predicate

        var error: NSError? = nil
        let count = self.managedObjectContext!.countForFetchRequest(fetchRequest, error: &error)
        
        return "\(range.location)-\(range.length) letter words: \(count)\n"
    }
    
    func endsWithGryWords() -> String {
        let text = NSExpression(forKeyPath: "text")
        let gry = NSExpression(forConstantValue: "gry")
        let predicate = NSComparisonPredicate(leftExpression: text, rightExpression: gry, modifier: .DirectPredicateModifier, type: .EndsWithPredicateOperatorType, options: nil)
        
        println("Predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = predicate
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let gryWords = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in gryWords {
                list.appendString(((word as NSManagedObject).valueForKey("text") as String)+",")
            }
        }
        return "-gry words: \(list)\n"
    }
    
    func anyWordContainsZ() -> String {
        let text = NSExpression(forKeyPath: "words.text")
        let z = NSExpression(forConstantValue: "z")
        let predicate = NSComparisonPredicate(leftExpression: text, rightExpression: z, modifier:.AnyPredicateModifier, type: .ContainsPredicateOperatorType, options: nil)
        
        println("Predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "WordCategory")
        fetchRequest.predicate = predicate
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let categories = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in categories {
                list.appendString(((word as NSManagedObject).valueForKey("firstLetter") as String)+",")
            }
        }
        
        return "ANY: \(list)\n"
    }
    
    func caseInsensitiveFetch(word: String) -> String {
        let text = NSExpression(forKeyPath: "text")
        let allCapsWord = NSExpression(forConstantValue: word.uppercaseString)
        let predicate = NSComparisonPredicate(leftExpression: text, rightExpression: allCapsWord, modifier: .DirectPredicateModifier, type: .EqualToPredicateOperatorType, options: .CaseInsensitivePredicateOption)
        
        println("Predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = predicate

        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            if words.isEmpty { return "" }
            else {
                let word: AnyObject! = words[0].valueForKey("text")
                return "\(word)\n"
            }
        }
        else { return "" }
    }
    
    func twentyLetterWordsEndingInIng() -> String {
        // Create a predicate that compares length to 20
        let length = NSExpression(forKeyPath: "length")
        let twenty = NSExpression(forConstantValue: 20)
        let predicateLength = NSComparisonPredicate(leftExpression: length, rightExpression: twenty, modifier: .DirectPredicateModifier, type: .EqualToPredicateOperatorType, options: nil)
        
        // Create a predicate that compares text to "ends with -ing"
        let text = NSExpression(forKeyPath: "text")
        let ing = NSExpression(forConstantValue: "ing")
        let predicateIng = NSComparisonPredicate(leftExpression: text, rightExpression: ing, modifier: .DirectPredicateModifier, type: .EndsWithPredicateOperatorType, options: nil)
        
        // Combine the predicates
        let predicate = NSCompoundPredicate.andPredicateWithSubpredicates([predicateLength, predicateIng])
        
        println("Compound predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.predicate = predicate
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in words {
                list.appendString(((word as NSManagedObject).valueForKey("text") as String)+",")
            }
        }
        
        return "\(list)\n"
    }
    
    func highCountCategories() -> String {
        let predicate = NSPredicate(format: "words.@count > %d", argumentArray: [10000])
        println("Predicate: \(predicate.predicateFormat)")
        
        var fetchRequest = NSFetchRequest(entityName: "WordCategory")
        fetchRequest.predicate = predicate
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let categories = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in categories {
                list.appendString(((word as NSManagedObject).valueForKey("firstLetter") as String)+",")
            }
        }
        
        return "High count categories: \(list)\n"
    }
    
    func averageWordLengthFromCollection() -> String {
        var fetchRequest = NSFetchRequest(entityName: "Word")
        
        var error: NSError? = nil
        let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error)
        
        var formattedAverage = "0"
        if let words = words {
            if let avg = (words as NSArray).valueForKeyPath("@avg.length") as? Float {
              let format = ".2"
              formattedAverage = avg.format(format)
            }
        }
        return "Average from collection: \(formattedAverage)\n"
    }
    
    func averageWordLengthFromExpressionDescription() -> String {
        let length = NSExpression(forKeyPath: "length")
        let average = NSExpression(forFunction: "average:", arguments: [length])
        
        var averageDescription = NSExpressionDescription()
        averageDescription.name = "average"
        averageDescription.expression = average
        averageDescription.expressionResultType = .FloatAttributeType
        
        var fetchRequest = NSFetchRequest(entityName: "Word")
        fetchRequest.propertiesToFetch = [averageDescription]
        fetchRequest.resultType = .DictionaryResultType
        
        var error: NSError? = nil
        let results = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error)
        
        var formattedAverage = "0"
        if let results = results {
            if let avg = (results[0] as NSDictionary).valueForKey("average") as? Float {
              let format = ".2"
              formattedAverage = avg.format(format)
            }
        }
        return "Average from expression description: \(formattedAverage)\n"
    }
    
    func longestWords() -> String {
        var fetchRequest = NSFetchRequest(entityName: "Word")
        let predicate = NSPredicate(format: "length = max:(length)", argumentArray: [])

        println("Predicate: \(predicate.predicateFormat)")
        
        fetchRequest.predicate = predicate        
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in words {
                list.appendString(((word as NSManagedObject).valueForKey("text") as String)+",")
            }
        }
        
        return "Longest words: \(list)\n"
    }
    
    func wordCategoriesWith25LetterWords() -> String {
        var fetchRequest = NSFetchRequest(entityName: "WordCategory")
        let predicate = NSPredicate(format: "SUBQUERY(words, $x, $x.length = 25).@count > 0", argumentArray: [])

        println("Subcategoy predicate: \(predicate.predicateFormat)")
        
        fetchRequest.predicate = predicate
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let categories = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in categories {
                list.appendString(((word as NSManagedObject).valueForKey("firstLetter") as String)+",")
            }
        }
        
        return "Categories with 25-letter words: \(list)\n"
    }
    
    func zWords() -> String {
        var fetchRequest = NSFetchRequest(entityName: "Word")
        let predicate = NSPredicate(format: "text BEGINSWITH 'z'", argumentArray: [])
        
        println("Predicate: \(predicate.predicateFormat)")
        
        let lengthSort = NSSortDescriptor(key: "length", ascending: true)
        let alphaSort = NSSortDescriptor(key: "text", ascending: false)
        
        fetchRequest.predicate = predicate
        fetchRequest.sortDescriptors = [lengthSort, alphaSort]
        
        let list = NSMutableString()
        var error: NSError? = nil
        if let words = self.managedObjectContext!.executeFetchRequest(fetchRequest, error: &error) {
            for word in words {
                list.appendString(((word as NSManagedObject).valueForKey("text") as String)+",")
            }
        }
        
        return "Z words: \(list)\n"
    }
}

extension Float {
    func format(f: String) -> String {
        return NSString(format: "%\(f)f", self)
    }
}